extensions/who-is-visiting-this-content/app.R

```diff
@@ -0,0 +1,151 @@
+        tabPanel(
+          "Aggregated Visits",
+          tableOutput("aggregated_visits")
+        )
+      )
+    )
+  )
+)
+
+server <- function(input, output, session) {
+  # Cache invalidation button ----
+  cache <- cachem::cache_disk("./app_cache/cache/")
+  observeEvent(input$clear_cache, {
+    print("Cache cleared!")
+    cache$reset()  # Clears all cached data
+    session$reload()  # Reload the app to ensure fresh data
+  })
+
+  observe({
+    if (nchar(input$content_guid) == 0) {
+      show("guid_input_msg")
+      hide("content_visit_tables")
+    } else {
+      hide("guid_input_msg")
+      show("content_visit_tables")
+    }
+  })
+
+  # Loading and processing data ----
+  client <- connect()
+
+  # Default dates. "This week" is best "common sense" best represented by six
+  # days ago thru the end of today. Without these, content takes too long to
+  # display on some servers.
+  date_range <- reactive({
+    list(
+      from_date = today() - ddays(6),
+      to_date = today()
+    )
+  })
+
+  content <- reactive({
+    # Somewhat hacky to get all content and filter for the guid, but makes error
+    # handling somewhat easier when printing content title.
+    get_content(client)
+  }) |> bindCache("static_key")
+
+  user_names <- reactive({
+    get_users(client) |>
+      mutate(full_name = paste(first_name, last_name)) |>
+      select(user_guid = guid, full_name, username)
+  }) |> bindCache("static_key")
+
+  usage_data <- reactive({
+    get_usage(
+      client,
+      from = date_range()$from_date,
+      to = date_range()$to_date + hours(23) + minutes(59) + seconds(59)
+    )
+  }) |> bindCache(date_range()$from_date, date_range()$to_date)
+
+  # Compute data
+  all_visits_data <- reactive({
+    usage_data() |>
+      filter(content_guid == input$content_guid) |>
+      left_join(user_names(), by = "user_guid") |>
+      replace_na(list(full_name = "[Anonymous]")) |>
+      arrange(desc(timestamp)) |>
+      select(timestamp, full_name, username)
+  }) |> bindCache(date_range()$from_date, date_range()$to_date, input$content_guid)
+
+  aggregated_visits_data <- reactive({
+    usage_data() |>
+      filter(content_guid == input$content_guid) |>
+      group_by(user_guid) |>
+      summarize(n_visits = n()) |>
+      left_join(user_names(), by = "user_guid") |>
+      replace_na(list(full_name = "[Anonymous]")) |>
+      arrange(desc(n_visits)) |>
+      select(n_visits, full_name, username)
+  }) |> bindCache(date_range()$from_date, date_range()$to_date, input$content_guid)
+
+  summary_message <- reactive({
+    content_title <- content() |>
+      filter(guid == input$content_guid) |>
+      pull(title)
+    hits <- all_visits_data()
+    glue(
+      "Content '{content_title}' had {nrow(hits)} between ",
+      "{min(hits$timestamp)} and {max(hits$timestamp)}."
+    )
+  })
+
+
+  output$summary_message <- renderText(summary_message())
+  output$all_visits <- renderTable(
+    all_visits_data()
+  )
```

Comment:
How hard would it be to add nice formatting to the timestamp column here?

Reply:
Very easy; its lack of presence is just a factor of "try to do a quick first draft". However, since we'll be switching the implementation to `gt` rather than `DT`, perhaps I won't implement it in `DT`?

Reply:
Sure, that's fine

Reply:
I have a note to try switching to `gt` in https://github.com/posit-dev/connect/issues/30322

Reply:
I added formatting. :)